src/map.js

```diff
@@ -38,82 +41,89 @@ class Cell {
 
     css() {
         if (this.p) return "player";
-        else if (this.i != null) return this.i.tt();
+        else if (this.i != null) {
+            if (this.i.tt == null) {
+                // Debugging code to show errors on map
+                console.log(this);
+                return "X";
+            }
+            else return this.i.tt();
+        }
         else return this.tt;
     }
 }
 
 
-function parseMap(d) {
+function parseMap(d, itemDefinitions) {
     console.log("Parsing");
+
     // Convert input text into array of arrays of characters (length 1 strings)
     var mapText = d3.dsvFormat("").parseRows(d).map(d => d[0].split('').map(d => d == " " ? "" : d));
 
-    //var maxLen = mapText.reduce((a, b) => Math.max(a, b.length), 0);
-
     var y = 0;
 
     // Convert strings into cell description objects
     const mapData = mapText.map(function (a) {
         var x = 0;
-        var row = a.map(function (c) {
-            var cell = new Cell(x, y);
-            var parseFn = parseDefault;
-
-            switch (c) {
-                case "@": parseFn = parsePlayer; break;
-                case "+": parseFn = parseDoor; break;
-                case "*": parseFn = parseGold; break;
-                case "¬": parseFn = parseKey; break;
-            }
-
-            parseFn(cell, c);
-
-            x++;
-
-            return cell;
-        });
-        y++; 
+        var row = a.map(c => parseCell(c, itemDefinitions, x++, y));
+        y++;
 
         return row;
     });
 
     // Flatten map for rendering
     const mapArray = mapData.reduce((a, b) => a.concat(b), []);
 
-    return {mapData, mapArray};
+    return { mapData, mapArray };
 
-    function parsePlayer(cell, c) {
-        parseDefault(cell, ".");
-        cell.p = true;
-    }
 
-    function parseDoor(cell, c) {
-        parseDefault(cell, ".");
-        cell.i = new door(cell.x%2 ? "red" : "green");
-    }
+// Turn each character into a cell, based on the symbol and accompanying 
+// definition in the json file.
+function parseCell(c, itemDefinitions, x, y) {
+    var cell = new Cell(x, y);
 
-    function parseGold(cell, c) {
-        parseDefault(cell, ".");
-        cell.i = new gold();
+    // Override tile symbol for parsing if player
+    if (c == "@") {
+        cell.p = true;
+        c = ".";
     }
 
-    function parseKey(cell, c) {
-        parseDefault(cell, ".");
-        // TODO(mstankiewicz): Sorry, testing hacks
-        cell.i = new key(cell.x%2 ? "red" : "green");
+    // Attempt to find this cell in item definition metadata
+    else if (itemDefinitions != null && itemDefinitions[c] != null) {
+        for (var i in itemDefinitions[c]) {
```

Comment:
I was assuming that (var i in array{}) would return array entries, rather than index -- is that right, or am I missing something important?

Reply:
Weirdly, it returns indices. Not sure why, hence the weird syntax.

Reply:
It was confusing to read.  Maybe a comment for the next reader?

Reply:
Use this instead:

```suggestion
        for (let item of itemDefinitions[c]) {
```
and do away with the declaration of `item` immediately following.

`for..in` enumerates keys (including array indices if used on arrays), while `for..of` enumerates values of iterable objects. `let` is irrelevant to this particular objection; it just behaves in a nicer, block-scoped way than function-scoped `var`.